tests/trainer/test_tp.py

```diff
@@ -1,27 +1,456 @@
+    )
+
+    return trainer
+
+
+def get_ddp_trainer(
+    size: int = 4,
+    batch_size: int = 1,
+    num_classes: int = 2,
+    num_features: int = 2,
+    seed: int = 44,
+    device: Union[torch.device, str] = 'cuda',
+    replication: int = 0,
+):
+    ddp_trainer = get_trainer(
+        size=size,
+        batch_size=batch_size,
+        num_classes=num_classes,
+        num_features=num_features,
+        seed=seed,
+        device=device,
+        replication=replication,
+    )
+    return ddp_trainer
+
+
+def get_fsdp_trainer(
+    size: int = 4,
+    batch_size: int = 1,
+    num_classes: int = 2,
+    num_features: int = 2,
+    seed: int = 44,
+    device: Union[torch.device, str] = 'cuda',
+    replication: int = 0,
+):
+    fsdp_config = FSDPConfig(
+        state_dict_type='full',
+        sharding_strategy='SHARD_GRAD_OP',
+        mixed_precision='full',
+        use_orig_params=True,
+    )
+    parallelism_config = ParallelismConfig(fsdp=fsdp_config)
+
+    fsdp_trainer = get_trainer(
+        parallelism_config=parallelism_config,
+        size=size,
+        batch_size=batch_size,
+        num_classes=num_classes,
+        num_features=num_features,
+        seed=seed,
+        device=device,
+        replication=replication,
+    )
+    return fsdp_trainer
+
+
+def get_tp_fsdp_trainer(
+    size: int = 4,
+    batch_size: int = 1,
+    num_classes: int = 2,
+    num_features: int = 2,
+    seed: int = 44,
+    device: Union[torch.device, str] = 'cuda',
+    replication: int = 0,
+):
+    from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel
+
+    fsdp_config = FSDPConfig(
+        state_dict_type='full',
+        sharding_strategy='SHARD_GRAD_OP',
+        mixed_precision='full',
+        use_orig_params=True,
+    )
+
+    layer_plan = {
+        'fc1': ColwiseParallel(),
+        'fc2': RowwiseParallel(),
+    }
+
+    tp_config = TPConfig(
+        layer_plan=layer_plan,
+        tensor_parallel_degree=replication,
+    )
+    parallelism_config = ParallelismConfig(fsdp=fsdp_config, tp=tp_config)
+
+    tp_fsdp_trainer = get_trainer(
+        parallelism_config=parallelism_config,
+        size=size,
+        batch_size=batch_size,
+        num_classes=num_classes,
+        num_features=num_features,
+        seed=seed,
+        device=device,
+        replication=replication,
+    )
+    return tp_fsdp_trainer
+
+
+def forward_pass(trainer):
+    reproducibility.seed_all(trainer.state.seed)
+    batch = next(iter(trainer.state.train_dataloader))
+    output = trainer.state.model.forward(batch)
+    return output
+
+
+def _replace_state_dict_name(state_dict: dict[str, Any], old_name: str, new_name: str) -> dict[str, Any]:
+    keys = list(state_dict.keys())
+    for key in keys:
+        if old_name in key:
+            new_key = key.replace(old_name, new_name, 1)
+            state_dict[new_key] = state_dict.pop(key)
+    return state_dict
+
+
+def compare_modules(
```

Comment:
isnt this implemented elsewhere in common?

Reply:
just found it - `deep_compare`

Reply:
I now use deep compare everywhere